src/libcollections/binary_heap.rs

```diff
@@ -1189,3 +1218,49 @@ impl<'a, T: 'a + Ord + Copy> Extend<&'a T> for BinaryHeap<T> {
         self.extend(iter.into_iter().cloned());
     }
 }
+
+#[unstable(feature = "collection_placement",
+           reason = "placement protocol is subject to change",
+           issue = "30172")]
+pub struct PlaceIn<'a, T: 'a>
+where T: Clone + Ord {
+    heap: &'a mut BinaryHeap<T>,
+}
+
+#[unstable(feature = "collection_placement",
+           reason = "placement protocol is subject to change",
+           issue = "30172")]
+impl<'a, T> Place<T> for PlaceIn<'a, T>
+where T: Clone + Ord {
+    fn pointer(&mut self) -> *mut T {
+        self.heap.data.place_back().pointer()
+    }
+}
+
+#[unstable(feature = "collection_placement",
+           reason = "placement protocol is subject to change",
+           issue = "30172")]
+impl<'a, T> Placer<T> for PlaceIn<'a, T>
+where T: Clone + Ord {
+    type Place = PlaceIn<'a, T>;
+
+    fn make_place(self) -> Self {
+        let _ = self.heap.data.place_back().make_place();
+        self
+    }
+}
+
+#[unstable(feature = "collection_placement",
+           reason = "placement protocol is subject to change",
+           issue = "30172")]
+impl<'a, T> InPlace<T> for PlaceIn<'a, T>
+where T: Clone + Ord {
+    type Owner = &'a T;
+
+    unsafe fn finalize(self) -> &'a T {
+        let len = self.heap.len();
+        let _ = self.heap.data.place_back().finalize();
+        let i = self.heap.sift_up_ind(0, len);
+        &mut self.heap.data[i]
```

Comment:
Returning an immutable reference seems sensible, but you needn’t to do a `&mut self.heap.data[i]` here; while return value of `finalize` cannot be reused, you can still avoid a bounds check by simply doing `self.heap.get_unchecked(i)`, as the index is guaranteed to be in bounds.